### PR TITLE
migration guide for no monadfail/st

### PR DESCRIPTION
**Pull request by ulysses4ever**
Cf. #33

### CODE REVIEW

File: guides/no-monadfail-st-inst.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# What
+
+CLC has approved the [proposal to remove the `ST` instance of the `MonadFail` class 
+from `base`][proposal]. `MonadFail` consists of the single method `fail`, which is
+meant to be implemented in terms of the monad itself. In contrast, the `ST` instance
+currently (GHC 9.2, `base-4.16.0.0`) have:
```
Comment (konsumlamm) [inline guides/no-monadfail-st-inst.md:6]:
```suggestion
currently (GHC 9.2, `base-4.16.0.0`) has:
```

File: guides/no-monadfail-st-inst.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+1. If you use `fail`, replace it with `error`.
+
+2. If you use a failable pattern, e.g.:
+
+```haskell
+  (x:xs) <- someFunc
+```
+
+you can replace it with a combination of `bind` and `let`, e.g.:
```
Comment (konsumlamm) [inline guides/no-monadfail-st-inst.md:43]:
```suggestion
you can replace it with a combination of `<-` and `let`, e.g.:
```
"bind" isn't actually an operatir, it's just how we call `>>=` and `<-` in do notation.

**Comment by andreasabel** [comment]
Good guide.

File: guides/no-monadfail-st-inst.md (hunk 2 above)
Comment (ulysses4ever) [reply guides/no-monadfail-st-inst.md]: Thank you ñ

File: guides/no-monadfail-st-inst.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+This goes against the above-mentioned idea behind the `fail` method. 
+
+[proposal]: https://github.com/haskell/core-libraries-committee/issues/33
+
+# When
+
+At the very earliest the change may appear in GHC 9.4 (~ Q3 2022), but
+given that a feature freeze for 9.4 is coming soon, it's more likely to
+happen in GHC 9.6 (~ Q2 2023).
```
Comment (Bodigrim) [inline guides/no-monadfail-st-inst.md:20]:
```suggestion
At the very earliest the change may appear in GHC 9.4 (~ Q3 2022).
```
Given that there are no prerequisites, I'd expect this to happen in 9.4.

File: guides/no-monadfail-st-inst.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+Here is a template, which you can use when raising PRs against affected
+libraries.
+
+> Title: Prepare for MonadFail/ST instance removal from base 
+>
+> CLC has approved the proposal to remove the `ST` instance of `MonadFail` from `base`
+> (https://github.com/haskell/core-libraries-committee/issues/33).
+>
+> The implementation of the proposal is delayed at least to GHC 9.4 and likely
+> to GHC 9.6, but one can already future-proof code to be
```
Comment (Bodigrim) [inline guides/no-monadfail-st-inst.md:66]:
```suggestion
> The implementation of the proposal is delayed to GHC 9.4,
> but one can already future-proof code to be
```

File: guides/no-monadfail-st-inst.md (hunk 3 above)
Comment (ulysses4ever) [reply guides/no-monadfail-st-inst.md]: Thanks!

**Review by Bodigrim — APPROVED** [review]

**Comment by Bodigrim** [comment]
Thanks @ulysses4ever!

**Pull request merged**